benchmarks/replay/replay.cpp

```diff
@@ -74,42 +79,128 @@ struct allocation {
  * @tparam MR The type of the `device_memory_resource` to use for allocation
  * replay
  */
-struct replay_benchmark {
+struct replay_benchmark : benchmark::Fixture {
+  MRFactoryFunc factory_;
   std::shared_ptr<rmm::mr::device_memory_resource> mr_{};
   std::vector<std::vector<rmm::detail::event>> const& events_{};
 
+  // Maps a pointer from the event log to an active allocation
+  std::unordered_map<uintptr_t, allocation> allocation_map;
+  std::mutex map_mutex;
+
+  // shared index to play back multithreaded events back in order
+  std::atomic_size_t event_index{0};
+
   /**
    * @brief Construct a `replay_benchmark` from a list of events and
    * set of arguments forwarded to the MR constructor.
    *
+   * @param factory A factory function to create the memory resource
    * @param events The set of allocation events to replay
    * @param args Variable number of arguments forward to the constructor of MR
    */
   replay_benchmark(MRFactoryFunc factory,
                    std::vector<std::vector<rmm::detail::event>> const& events)
-    : mr_{factory()}, events_{events}
+    : factory_{factory}, mr_{}, events_{events}, allocation_map{events.size()}, event_index{0}
+  {
+  }
+
+  /**
+   * @brief Copy construct a replay_benchmark
+   *
+   * Does not copy the mutex or the map
+   */
+  replay_benchmark(replay_benchmark const& other)
+    : factory_{other.factory_},
+      mr_{other.mr_},
+      events_{other.events_},
+      allocation_map{events_.size()},
+      event_index{0}
+  {
+  }
+
+  /// Add an allocation to the map (thread safe)
+  void set_allocation(std::unordered_map<uintptr_t, allocation>& allocation_map,
+                      uintptr_t ptr,
+                      allocation alloc)
+  {
+    std::lock_guard<std::mutex> lock(map_mutex);
+    allocation_map.insert({ptr, alloc});
+  }
+
+  /// Remove an allocation from the map (thread safe)
+  allocation remove_allocation(std::unordered_map<uintptr_t, allocation>& allocation_map,
+                               uintptr_t ptr)
   {
+    std::lock_guard<std::mutex> lock(map_mutex);
+    auto iter = allocation_map.find(ptr);
+    if (iter != allocation_map.end()) {
+      allocation a = iter->second;
+      allocation_map.erase(iter);
+      return a;
+    }
+    return allocation{};
+  }
+
+  /// Create the memory resource shared by all threads before the benchmark runs
+  void SetUp(const ::benchmark::State& state) override
+  {
+    if (state.thread_index == 0) {
+      rmm::logger().log(spdlog::level::info, "------ Start of Benchmark -----");
+      mr_ = factory_();
+    }
```

Comment:
Why not just put this logic in the `replay_benchmark` ctor? That way you don't have to do the check against the `thread_index`. Only one instance of the fixture object is constructed and shared by all threads.

Reply:
THe problem is the replay_benchmarks ctor is called when you register the benchmark, and the dtor is not called until after ALL the benchmarks have run. So you end up constructing all the MRs you are benchmarking first, then running the benchmarks, then destroying them. If you have a pool_mr and a binning_mr, then you effectively use up all the GPU memory before any benchmark runs, and the result is OOM during the first benchmark of one of those MRs.  This is the whole reason for all the shenanigans with the fixture. 

Google benchmark is not well designed for this case. It took me the better part of a day of reading their undocumented code to figure out a workaround.

Reply:
Also each benchmark may be run multiple times, and you don't want to reuse the pool (or time creation of the pool) across runs.